cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataReaderWriterModules.java

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright © 2018 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.metadata;
+
+import co.cask.cdap.api.metadata.MetadataReader;
+import co.cask.cdap.api.metadata.MetadataWriter;
+import co.cask.cdap.common.metadata.AbstractMetadataClient;
+import co.cask.cdap.common.runtime.RuntimeModule;
+import com.google.inject.AbstractModule;
+import com.google.inject.Module;
+
+/**
+ * Guice binding for {@link MetadataReader} and {@link MetadataWriter} for service binding see
+ * {@link MetadataServiceModule}.
+ */
+public class MetadataReaderWriterModules extends RuntimeModule {
+
+  @Override
+  public Module getInMemoryModules() {
+    return new AbstractModule() {
+      @Override
+      protected void configure() {
+        bind(MetadataReader.class).to(DefaultMetadataReader.class);
+      }
+    };
+  }
+
+  @Override
+  public Module getStandaloneModules() {
+    return new AbstractModule() {
+      @Override
+      protected void configure() {
+        bind(MetadataReader.class).to(DefaultMetadataReader.class);
+      }
+    };
+  }
+
+  @Override
+  public Module getDistributedModules() {
+    return new AbstractModule() {
+      @Override
+      protected void configure() {
+        // CDAP-13610 We should be binding MetadataAdmin which the AbstractMetadataAdmin implements.
```

Comment:
I am not sure I understand this comment... why do the in-memory and local modules not bind this, and where is this used? Do we inject a AbstractMetadataClient anywhere?

Reply:
We need two implementation of the MetadataClient (rest based and another remote which can do discovery). Currently we don't have any interface for metadata client its just one gigantic class. All the logic reside in AbstractMetadataClient and there is a MetadataClient which extends it and uses restClient serve the request. The RemoteMetadataClient which is introduced in this PR uses a remoteClient and service discovery to serve the request.

AbstractMetadataClient is injected in RemoteMetadataReader to support the calls made from programatic apis. In case of local and standalone mode the DefaultMetadataReader is used. Since neither of these mode is distributed and doesn't have impersonation DefaultMetadataReader directly reads the store through MetadataAdmin rather than going through a client.